openshift-with-appstudio-test/e2e/jvm_build_service_test.go

```diff
@@ -265,4 +268,97 @@ func TestExampleRun(t *testing.T) {
 			debugAndFailTest(ta, "timed out waiting for shaded-jdk11 to complete")
 		}
 	})
+
+	ta.t.Run("make sure second build access cached dependencies", func(t *testing.T) {
+		defer dumpBadEvents(ta)
+		defer dumpPods(ta, "jvm-build-service")
+		ta.run = &v1beta1.PipelineRun{}
+		obj = streamFileYamlToTektonObj(runYamlPath, ta.run, ta)
+		ta.run, ok = obj.(*v1beta1.PipelineRun)
+		if !ok {
+			debugAndFailTest(ta, fmt.Sprintf("file %s did not produce a pipelinerun: %#v", runYamlPath, obj))
+		}
+		ta.run, err = tektonClient.TektonV1beta1().PipelineRuns(ta.ns).Create(context.TODO(), ta.run, metav1.CreateOptions{})
+		if err != nil {
+			debugAndFailTest(ta, err.Error())
+		}
+
+		ctx := context.TODO()
+		watch, werr := tektonClient.TektonV1beta1().PipelineRuns(ta.ns).Watch(ctx, metav1.ListOptions{})
+		if werr != nil {
+			debugAndFailTest(ta, fmt.Sprintf("error creating watch %s", werr.Error()))
+		}
+		defer watch.Stop()
+
+		exitForLoop := false
+		podClient := kubeClient.CoreV1().Pods(ta.ns)
+
+		for {
+			select {
+			case <-ctx.Done():
+				ta.Logf("context done")
+				exitForLoop = true
+				break
+			case <-time.After(15 * time.Minute):
+				ta.Logf("timed out waiting for second build to complete")
+				exitForLoop = true
+				break
+			case event := <-watch.ResultChan():
```

Comment:
Memory requirements is the con/downside of watches.  Removal of polling gaps, etc is the pro/upside.

In fact, with the periodic, I backed off watches a bit (to only watch dependencybuilds and not both artifactbuilds and dependencybuilds) since the service registry build created so many, and on a couple of occassions I saw the test process freeze up with a huge garbage collection event / sometimes crashed.

Sam concern with controllers.  We have to be careful if we reconcile on a type / set of namespaces that are too onerous.